File: treeshaders_2_cpp/agent.cs
using Godot;
using System;
using System.ComponentModel;

public partial class agent : Node3D
{

	private NavigationAgent3D navigation_agent;
	private AnimationPlayer anim_player;
	private Node3D trees;
	public bool activated = false;

	//[Signal]
	//public delegate void pathEventHandler(agent arg1);


	public override void _Ready()
	{
		navigation_agent = GetNode<NavigationAgent3D>("NavigationAgent3D");
		anim_player = GetNode<AnimationPlayer>("AnimationPlayer");
		trees = GetNode<Node3D>("../../NavigationRegion3D/trees");
		navigation_agent.PathDesiredDistance = 0.5f;
		navigation_agent.TargetDesiredDistance = 0.5f;
		anim_player.Play("walking");
	}

	public override void _Process(double delta)
	{
		if (activated && navigation_agent.IsNavigationFinished()) {
			// EmitSignal(SignalName.path, (agent)this);
			give_agent_path();
			return;
		}

		Vector3 current_agent_position = this.GlobalPosition;
		Vector3 next_path_position = navigation_agent.GetNextPathPosition();

		Vector3 new_velocity = next_path_position - current_agent_position;
		new_velocity = new_velocity.Normalized();
		new_velocity = new_velocity * 0.25f;

		Vector3 opposite_direction = current_agent_position - next_path_position;
		Vector3 target_position_for_reverse_look = current_agent_position + opposite_direction.Normalized();

		LookAt(target_position_for_reverse_look, new Vector3(0.01f, 1f, 0.01f));

		Position += new Vector3(new_velocity.X * (float)delta, new_velocity.Y * (float)delta, new_velocity.Z * (float)delta);
	}


	public void init(int i)
	{
		GetNode<Area3D>("Area3D").Name = i.ToString();
	}

	private void give_agent_path()
	{
		Godot.Collections.Array<Godot.Node> targets = trees.GetChildren();
		Random rnd = new Random();
		Node3D target = (Node3D)targets[rnd.Next(targets.Count)];
		navigation_agent.TargetPosition = target.GlobalPosition;
	}
}

File: src/agent.cpp
#include "agent.h"

#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/classes/engine.hpp>

using namespace godot;

void Agent::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_activated"), &Agent::get_activated);
    ClassDB::bind_method(D_METHOD("set_activated", "p_activated"), &Agent::set_activated);
    ClassDB::add_property("Agent", PropertyInfo(Variant::BOOL, "activated"), "set_activated", "get_activated");

    ADD_SIGNAL(MethodInfo("path", PropertyInfo(Variant::OBJECT, "node")));

    ClassDB::bind_method(D_METHOD("init"), &Agent::init);
    ClassDB::bind_method(D_METHOD("goto_target", "node"), &Agent::goto_target);
}

Agent::Agent() {
    // Initialize any variables here.
}

Agent::~Agent() {
    // Add your cleanup here.
}

void Agent::_process(double delta) {
    if (Engine::get_singleton()->is_editor_hint()) {
        return;
    }

    if (activated && navigation_agent->is_navigation_finished()) {
        emit_signal("path", this);
    }

    Vector3 next_path_position = navigation_agent->get_next_path_position();

    if (next_path_position == last_path_position) {
        return;
    }

    Vector3 new_velocity = next_path_position - get_global_position();
    new_velocity = new_velocity.normalized() * 0.25;

    Vector3 opposite_direction = get_global_position() - next_path_position;
    Vector3 target_position_for_reverse_look = get_global_position() + opposite_direction.normalized();

    look_at(target_position_for_reverse_look, Vector3(0.01, 1, 0.01));

    set_position(get_position() + new_velocity * delta);

    last_path_position = get_global_position();
}

void Agent::_ready() {
    if (Engine::get_singleton()->is_editor_hint()) {
        return;
    }
    navigation_agent = get_node<NavigationAgent3D>("NavigationAgent3D");
    navigation_agent->set_path_desired_distance(0.5);
    navigation_agent->set_target_desired_distance(0.5);
    anim_player = get_node<AnimationPlayer>("AnimationPlayer");
    anim_player->play("walking");
}

void Agent::set_activated(const double p_activated) {
    activated = p_activated;
}

bool Agent::get_activated() const {
    return activated;
}

void Agent::goto_target(Node3D* node) {
//void Agent::goto_target(const Vector3& target) {
    navigation_agent->set_target_position(node->get_global_position());
    //navigation_agent->set_target_position(target);
}

void Agent::init(int i) {
    uniq = i;
}